src/Catalog/Icons/IconsCollector.cs

```diff
@@ -0,0 +1,121 @@
+        }
+
+        protected override Task<IEnumerable<CatalogCommitItemBatch>> CreateBatchesAsync(
+            IEnumerable<CatalogCommitItem> catalogItems)
+        {
+            var maxCommitTimestamp = catalogItems.Max(x => x.CommitTimeStamp);
+
+            return Task.FromResult<IEnumerable<CatalogCommitItemBatch>>(new[]
+            {
+                new CatalogCommitItemBatch(
+                    catalogItems,
+                    key: null,
+                    commitTimestamp: maxCommitTimestamp),
+            });
+        }
+
+        protected override async Task<bool> OnProcessBatchAsync(
+            CollectorHttpClient client,
+            IEnumerable<CatalogCommitItem> items,
+            JToken context,
+            DateTime commitTimeStamp,
+            bool isLastBatch,
+            CancellationToken cancellationToken)
+        {
+            await _iconCopyResultCache.InitializeAsync(cancellationToken);
+
+            var filteredItems = items
+                .GroupBy(i => i.PackageIdentity)                          // if we have multiple commits for the same package (id AND version)
+                .Select(g => g.OrderBy(i => i.CommitTimeStamp).ToList()); // group them together for processing in order
+            var itemsToProcess = new ConcurrentBag<IReadOnlyCollection<CatalogCommitItem>>(filteredItems);
+            var tasks = Enumerable
+                .Range(1, ServicePointManager.DefaultConnectionLimit)
+                .Select(_ => ProcessIconsAsync(itemsToProcess, cancellationToken));
+            await Task.WhenAll(tasks);
+
+            await _iconCopyResultCache.SaveAsync(cancellationToken);
+
+            return true;
+        }
+
+        private async Task ProcessIconsAsync(
+            ConcurrentBag<IReadOnlyCollection<CatalogCommitItem>> items,
+            CancellationToken cancellationToken)
+        {
+            await Task.Yield();
+            var storage = _targetStorageFactory.Create();
+
+            using (_logger.BeginScope("{CallGuid}", Guid.NewGuid()))
+            while (items.TryTake(out var entries))
+            {
+                var firstItem = entries.First();
+                using (_logger.BeginScope("Processing commits for {PackageId} {PackageVersion}", firstItem.PackageIdentity.Id, firstItem.PackageIdentity.Version))
+                {
+                    foreach (var item in entries)
+                    {
+                        if (item.IsPackageDetails)
+                        {
+                            PackageDetailsCatalogLeaf leaf;
+                            try
+                            {
+                                leaf = await _catalogClient.GetPackageDetailsLeafAsync(item.Uri.AbsoluteUri);
+                            }
+                            catch (Exception e)
+                            {
+                                _logger.LogError(0, e, "Error while trying to retrieve catalog leaf {LeafUrl}", item.Uri.AbsoluteUri);
```

Comment:
>LogError [](start = 40, length = 8)

Won't this log twice? One here and one at the entry point? If we do this how many times does the exception appear in:

- traces
- exceptions

Reply:
Yes, it will. But this one has proper logging scope and leaf URL in the message. The one at the entry point is utterly useless for this type of failure as it has none and we have 100 parallel tasks running, so even guessing the URL from logs does not work.